### PR TITLE
Automatically add active class to slotted element in Link component

### PR DESCRIPTION
**Pull request by scottmcclung**
### Active attribute

Implemented reflection of an "active" attribute on the Link component when the link is active.  This allows for the parent to apply styles to any content in the Link slot of the active links.

For example, if the currentPath was "/about" the `About` anchor tag slotted in the Link component would be bold. 
  
exampleComponent.html
```html
<c-router>
  <c-link to="/about" active>
    <a>About</a>
  </c-link>
  <c-link to="/contact">
    <a>Contact</a>
  </c-link>
</c-router>
```

exampleComponent.css
```css
c-link[active] a {
  font-weight: bold;
}
```

### Automatic active class assignment

Implemented automatic active class assignment to the top level element slotted in the Link component when the link is active.  By default the class name "link-active-class" is applied.

For example, if the currentPath is "/about" the content slotted in the about Link component will have link-active-class automatically added to it.
  
exampleComponent.html
```html
<c-router>
  <c-link to="/about" active>
    <a class="link-active-class">About</a>
  </c-link>
  <c-link to="/contact" active>
    <a>Contact</a>
  </c-link>
</c-router>
```

exampleComponent.css
```css
.link-active-class {
  color: red;
}
```

In the case that we want to apply specific classes (e.g. SLDS classes) to the slotted content, we can set the "active-class" attribute on the Link component.  The provided class name is automatically added to the content in the slot.

exampleComponent.html
```html
<c-router>
  <c-link to="/about" active-class="slds-text-title_caps" active>
    <a class="slds-text-title_caps">About</a>
  </c-link>
  <c-link to="/contact" active-class="slds-text-title_caps">
    <a>Contact Us</a>
  </c-link>
</c-router>
```

The activeClass prop can also accept multiple classes separated by spaces.

exampleComponent.html
```html
<c-router>
  <c-link to="/about" active-class="slds-text-title_caps slds-truncate" active>
    <a class="slds-text-title_caps slds-truncate">About</a>
  </c-link>
  <c-link to="/contact" active-class="slds-text-title_caps slds-truncate">
    <a>Contact</a>
  </c-link>
</c-router>
```

The active class will also be updated when the content in the slot is changed.  For example, if we have conditionally rendered content, the activeClass would be re-applied when the condition changes. 

exampleComponent.html
```html
<c-router>
  <c-link to="/about" active-class="slds-text-title_caps slds-truncate" active>
    <template if:true={showMyProfile}>
      <a class="slds-text-title_caps slds-truncate">My Profile</a>
    </template>
    <template if:false={showMyProfile}>
      <a class="slds-text-title_caps slds-truncate">About</a>
    </template>
  </c-link>
</c-router>
```

### CODE REVIEW

File: force-app/main/default/lwc/link/link.html [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <template>
-    <slot onclick={handleClick}>
+    <slot onclick={handleClick} onslotchange={handleSlotChange}>
```
Comment (chandrakiran-dev) [inline force-app/main/default/lwc/link/link.html:2]: I have install this on my org, but for me `onslotchange` event not calling after changing the Link.

image

In the above image, On the load its properly setting the active attribute to 'item one' but after clicking on 'item two', Active attribute should change to Link 'item two'.

Please review at your end.

Thanks :)

File: force-app/main/default/lwc/link/link.js [hunk 2]
```diff
@@ -35,20 +47,33 @@ export default class Link extends LightningElement {
...
+    }
     handleClick(e) {
         e.stopPropagation();
         if (this.to) {
             this.routerInstance.currentPath = this.to;
             this.currentPath = this.to;
         }
     }
+    renderedCallback() {
+        this._slottedElement = this.querySelector('*');
```
Comment (chandrakiran-dev) [inline force-app/main/default/lwc/link/link.js]: In my case, `this._slottedElement` always blank.

image

**Review by chandrakiran-dev — CHANGES_REQUESTED** [review]

File: force-app/main/default/lwc/link/link.html (hunk 1 above)
Comment (scottmcclung) [reply force-app/main/default/lwc/link/link.html:2]: Good catch!   Forgot to set the active flag in the route subscription callback ( this.setCurrentPath() ).
That ensures that the active state is updated when any link causes the navigation event.
Fixed.

File: force-app/main/default/lwc/link/link.js (hunk 2 above)
Comment (scottmcclung) [reply force-app/main/default/lwc/link/link.js]: This brings up a good question for you.  
This PR only sets the active class on content that is slotted into the link component, but not on the default lightning-button.  Given that the lightning-button component is very opinionated and is styled with the variant prop it seemed like it was less useful to set a class on it.  So _slottedElement will only have a value if theres something inserted in the slot.
What behavior do you think is correct here?  Should we have the active classes applied to the default lightning-button component?

File: force-app/main/default/lwc/link/link.html (hunk 1 above)
Comment (chandrakiran-dev) [reply force-app/main/default/lwc/link/link.html:2]: Awesome :) It's working now very well.

File: force-app/main/default/lwc/link/link.js (hunk 2 above)
Comment (chandrakiran-dev) [reply force-app/main/default/lwc/link/link.js]: You are right, It's working perfectly with the slotted into the link component. 
I thing, It's fine for default one.

Changes looks awesome. Thanks

**Review by chandrakiran-dev — APPROVED** [review]

**Pull request merged**